synthesizer/src/vm/helpers/history.rs

```diff
@@ -0,0 +1,123 @@
+        StorageMode::Production | StorageMode::Custom(_) => format!("{HISTORY_DIRECTORY_NAME}-{network}"),
+    };
+
+    // Obtain the path to the ledger.
+    let mut path = aleo_ledger_dir(network, storage_mode.clone());
+    // Go to the folder right above the ledger.
+    path.pop();
+    // Append the history directory's name.
+    path.push(directory_name);
+
+    path
+}
+
+#[derive(Copy, Clone, Serialize, Deserialize)]
+#[serde(rename_all = "lowercase")]
+pub enum MappingName {
+    /// The `bonded` mapping.
+    Bonded,
+    /// The `delegated` mapping.
+    Delegated,
+    /// The `unbonding` mapping.
+    Unbonding,
+}
+
+impl Display for MappingName {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::Bonded => write!(f, "bonded"),
+            Self::Delegated => write!(f, "delegated"),
+            Self::Unbonding => write!(f, "unbonding"),
+        }
+    }
+}
+
+pub struct History {
+    /// The path to the history directory.
+    path: PathBuf,
+}
+
+impl History {
+    /// Initializes a new instance of `History`.
+    pub fn new(network: u16, storage_mode: &StorageMode) -> Self {
+        Self { path: history_directory_path(network, storage_mode) }
+    }
+
+    /// Stores a mapping from a given block in the history directory as JSON.
+    pub fn store_mapping<T>(&self, height: u32, mapping: MappingName, data: &T) -> Result<()>
+    where
+        T: Serialize + ?Sized,
+    {
+        // Get the path to the block directory.
+        let path = self.block_path(height);
+        // Create the block directory if it does not exist.
+        if !path.exists() {
+            std::fs::create_dir_all(&path)?;
+        }
+
+        // Write the entry to the block directory.
+        let path = path.join(format!("block-{height}-{mapping}.json"));
+        std::fs::write(path, serde_json::to_string_pretty(data)?)?;
+
+        Ok(())
+    }
+
+    /// Loads the JSON string for a mapping from a given block from the history directory.
+    pub fn load_mapping(&self, height: u32, mapping: MappingName) -> Result<String> {
+        // Get the path to the block directory.
+        let path = self.block_path(height);
+        // Get the path to the block file.
+        let path = path.join(format!("block-{height}-{mapping}.json"));
+
+        // Read the file.
+        let data = std::fs::read_to_string(path)?;
+
+        Ok(data)
+    }
+
+    // A helper function to get the path to the block directory.
+    fn block_path(&self, height: u32) -> PathBuf {
+        // Get the path the directory group.
+        let group = Self::group(height);
+        let path = self.path.join(format!("group-{group}"));
+        // Get the path to the block directory.
+        path.join(format!("block-{height}"))
+    }
+
+    // A helper function to calculate the group number for a given block height.
+    fn group(height: u32) -> u32 {
+        height.saturating_div(2u32 << 16)
```

Comment:
```suggestion
        height.saturating_div(1u32 << 16)
```
Or perhaps even clearer, use: `u16::MAX`